src/world/area_omo/omo_02/D9D510.c

```diff
@@ -51,9 +53,161 @@ ApiStatus func_80240040_D9D510(Evt* script, s32 isInitialCall) {
     return ApiStatus_BLOCK;
 }
 
-INCLUDE_ASM(s32, "world/area_omo/omo_02/D9D510", func_802402C0_D9D790);
+void func_802402C0_D9D790(Gfx* displayList, s32* outDist) {
+    char pad[0x8];
+    Gfx* dlist = displayList;
+    u8 cmd;
 
-INCLUDE_ASM(s32, "world/area_omo/omo_02/D9D510", func_8024043C_D9D90C);
+    do {
+        u32 temp_s2 = dlist->words.w0;
+        u32 temp_s3 = dlist->words.w1;
+
+        cmd = temp_s2 >> 0x18;
+        if (cmd == G_DL) {
+            func_802402C0_D9D790((Gfx*) temp_s3, outDist);
+        }
+        if (cmd == G_VTX) {
+            s32 numVertices = (temp_s2 >> 0xC) & 0xFF;
+            Vtx* vtx = (Vtx*) temp_s3;
+            f32 subX, subY, subZ;
+            s32 i;
+
+            subX = subY = subZ = 0.0f;
+
+            *outDist = -1;
+
+            for (i = 0; i < numVertices; i++) {
+                f32 x = vtx[i].v.ob[0] - subX;
+                f32 y = vtx[i].v.ob[1] - subY;
+                f32 z = vtx[i].v.ob[2] - subZ;
+                f32 dist = sqrtf(SQ(x) + SQ(y) + SQ(z));
+
+                if (*outDist < dist) {
+                    *outDist = dist;
+                }
+            }
+        }
+        dlist++;
+    } while (cmd != G_ENDDL);
+}
+
+typedef struct UnkOmoThing {
+    /* 0x00 */ s32 unk_00;
+    /* 0x04 */ Vec3f center;
+    /* 0x10 */ Vec3f unk_10;
+    /* 0x1C */ Vec3f rot;
+    /* 0x28 */ Vec3f unk_28;
+    /* 0x34 */ f32 unk_34;
+    /* 0x38 */ f32 unk_38;
+    /* 0x3C */ f32 unk_3C;
+    /* 0x40 */ s32 treeIndex;
+    /* 0x44 */ s32 colliderID;
+    /* 0x48 */ Matrix4f transformMatrix;
+    /* 0x88 */ s32 unk_88;
+} UnkOmoThing; // size = 0x8C
+
+ApiStatus func_8024043C_D9D90C(Evt* script, s32 isInitialCall) {
+    Matrix4f sp18, sp58, sp98, spD8;
+    Model* model;
+    UnkOmoThing* data;
+    UnkOmoThing* it;
+    u32 i;
+    s32 j, k;
+
+    if (isInitialCall) {
+        script->functionTempPtr[0] = heap_malloc(sizeof(*data) * 12);
+        script->functionTemp[1] = 0;
+
+        it = (UnkOmoThing*) script->functionTempPtr[0];
+        for (i = 0; i < 12; i++, it++) {
```

Comment:
The 12s here, on L153, and L205 should probably be parameterised - perhaps `ARRAY_COUNT(D_80243580_DA0A50)`?